### PR TITLE
sshd: separate key generation into another service

### PR DESCRIPTION
**Pull request by nhooyr**
###### Motivation for this change

~~Quick and dirty~~ fix for #19589.

~~In the future, I think we should use a separate key generation service.~~ (done)
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/networking/ssh/sshd.nix [hunk 1]
```diff
@@ -272,6 +274,7 @@ in
                 KillMode = "process";
               } // (if cfg.startWhenNeeded then {
                 StandardInput = "socket";
+                StandardError = "journal";
```
Comment (Mic92) [inline nixos/modules/services/networking/ssh/sshd.nix]: Does this also suppress other error message, which should  be not redirected?

**Comment by Mic92** [comment]
Rather then the quickfix, I would directly make it right.
This is how archlinux solve this problem:

```
# /usr/lib/systemd/system/sshdgenkeys.service
[Unit]
Description=SSH Key Generation
ConditionPathExists=|!/etc/ssh/ssh_host_dsa_key
ConditionPathExists=|!/etc/ssh/ssh_host_dsa_key.pub
ConditionPathExists=|!/etc/ssh/ssh_host_ecdsa_key
ConditionPathExists=|!/etc/ssh/ssh_host_ecdsa_key.pub
ConditionPathExists=|!/etc/ssh/ssh_host_ed25519_key
ConditionPathExists=|!/etc/ssh/ssh_host_ed25519_key.pub
ConditionPathExists=|!/etc/ssh/ssh_host_rsa_key
ConditionPathExists=|!/etc/ssh/ssh_host_rsa_key.pub

[Service]
ExecStart=/usr/bin/ssh-keygen -A
Type=oneshot
RemainAfterExit=yes
```

```
# /usr/lib/systemd/system/sshd.service
[Unit]
Description=OpenSSH Daemon
Wants=sshdgenkeys.service
After=sshdgenkeys.service
After=network.target

[Service]
ExecStart=/usr/bin/sshd -D
ExecReload=/bin/kill -HUP $MAINPID
KillMode=process
Restart=always

[Install]
WantedBy=multi-user.target
```

**Comment by peterhoeg** [comment]
As a matter of principle, I think we should avoid creating DSA and ECDSA keys or at least have checking in place to only generate it for old version similarly to how it is done in ```nixos/modules/services/networking/ssh/sshd.nix```:

``` nix
supportOldHostKeys = !versionAtLeast config.system.stateVersion "15.07";
```

**Comment by Mic92** [comment]
@peterhoeg is something wrong with ecdsa?

**Comment by nhooyr** [comment]
@Mic92 If I remember correctly, any client that supports ECDSA also supports ED25519 (ED25519 is normally preferred over ECDSA) and any client that supports DSA also supports RSA (RSA is normally preferred over DSA).

**Comment by nhooyr** [comment]
Ok, done.

**Comment by Mic92** [comment]
thanks!

**Pull request merged**